// @andromedaprotocol/cli/dist/handlers/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.allCommands = exports.userHandler = exports.txHandler = exports.gqlHandler = exports.bankHandler = exports.adoHandler = exports.chainHandler = exports.wasmHandler = exports.walletHandler = exports.vfs = exports.tx = exports.bank = exports.gql = exports.ado = exports.chain = exports.wasm = exports.wallets = void 0;
const chain_1 = __importDefault(require("./chain"));
const wallet_1 = __importDefault(require("./wallet"));
const wasm_1 = __importDefault(require("./wasm"));
const ado_1 = __importDefault(require("./ado"));
const gql_1 = __importDefault(require("./gql"));
const bank_1 = __importDefault(require("./bank"));
const tx_1 = __importDefault(require("./tx"));
const user_1 = __importDefault(require("./user"));
const utils_1 = require("./utils");
exports.wallets = __importStar(require("./wallet"));
exports.wasm = __importStar(require("./wasm"));
exports.chain = __importStar(require("./chain"));
exports.ado = __importStar(require("./ado"));
exports.gql = __importStar(require("./gql"));
exports.bank = __importStar(require("./bank"));
exports.tx = __importStar(require("./tx"));
exports.vfs = __importStar(require("./user"));
__exportStar(require("./utils"), exports);
exports.walletHandler = (0, utils_1.generateHandler)(wallet_1.default, "wallet");
exports.wasmHandler = (0, utils_1.generateHandler)(wasm_1.default, "wasm");
exports.chainHandler = (0, utils_1.generateHandler)(chain_1.default, "chain");
exports.adoHandler = (0, utils_1.generateHandler)(ado_1.default, "ado");
exports.bankHandler = (0, utils_1.generateHandler)(bank_1.default, "bank");
exports.gqlHandler = (0, utils_1.generateHandler)(gql_1.default, "gql");
exports.txHandler = (0, utils_1.generateHandler)(tx_1.default, "tx");
exports.userHandler = (0, utils_1.generateHandler)(user_1.default, "user");
exports.allCommands = [
    ...Object.keys(chain_1.default).map((cmd) => `chain ${cmd}`),
    ...Object.keys(wallet_1.default).map((cmd) => `wallet ${cmd}`),
    ...Object.keys(wasm_1.default).map((cmd) => `wasm ${cmd}`),
    ...Object.keys(tx_1.default).map((cmd) => `tx ${cmd}`),
    ...Object.keys(ado_1.default).map((cmd) => `ado ${cmd}`),
    ...Object.keys(gql_1.default).map((cmd) => `gql ${cmd}`),
    ...Object.keys(bank_1.default).map((cmd) => `bank ${cmd}`),
    ...Object.keys(user_1.default).map((cmd) => `user ${cmd}`),
];
